rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -150,3 +157,74 @@ impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
         internal_server_error()
     }
 }
+
+#[derive(Clone)]
+pub struct ServerRequestIdResponseProvider<S> {
+    inner: S,
+    header_key: HeaderName,
+}
+
+/// A layer that inserts the generated [`ServerRequestId`] to the response headers
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct ServerRequestIdResponseProviderLayer {
+    header_key: Box<str>,
```

Comment:
I think `header_key` should be a `HeaderName`. Similarly the  constructor should accept `HeaderName` - the reason being is `HeaderName` has a `const from_static` method which avoids `Box<str>` entirely. It also means that `Layer::layer` becomes infallible.